Tms570 support

kamathba opened the pull request:
Review patchset

ndreys committed:
hw/char: Add support for Hercules RTP module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
target/arm: Change bswap_code()'s parameter to CPUARMState *

Change bswap_code()'s parameter to CPUARMState * in order to accomodate
upcoming changes. No functional change intended.

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
target/arm: Add support for SCTLR.IE used in ARMv7-R

ARMv7-R can support legacy BE32 code, which can be configured by SoC
vendor via SCTLR.IE bit. Add various bit and pieces needed to emulate
that feature.

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
target/arm: gdbstub: Swap register endiannes if SCTLR.IE is set

We need to swap endiannes of register value data sent to GDB when
SCTLR.IE is set, since GDB will be running in BE32 and all of the
instruction memory is going to be big-endian, so it'll expect register
data to be as well.

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add support for Hercules System module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/intc: Add support for Hercules VIM module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/gpio: Add support for Hercules GIO/N2HET blocks

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/timer: Add support for Hercules RTI module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/net: Add support for emulating Hercules EMAC

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/ssi: Add support for Hercules MiB SPI

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/adc: Add support for Hercules MiBADC

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/arm: Add support for Hercules SoC

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/dma: Add support for Hercules DMA

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add support for Hercules SCM module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add support for Hercules eFUSE module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add support for Hercules L2RAMW module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add support for Hercules PMM module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add emulation for Hercules STC block

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add emulation for Hercules PBIST block

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add support for Hercules CCM IP block

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add support for Hercules L2FMC IP block

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
bitmap: Add for_each_set_bit macro

Add for_each_set_bit macro as found in Linux kernel

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

ndreys committed:
hw/misc: Add emulation for Hercules ECAP IP block

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

Commit:
MAINTAINERS: add entry for TI Hercules chip family

Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

kamathba commented on hw/net/hercules_emac.c:
```diff
@@ -0,0 +1,650 @@
+    uint32_t next;
+
+    uint32_t buffer_pointer;
+#ifdef HOST_WORDS_BIGENDIAN
```
The only reason I could see why this is needed would be something to do with QEMUs DMA type, but even that would make me a little concerned. On the target, it seems like buffer_length should immediately follow the buffer pointer.

ndreys replied:
As discussed, let's nuke it from the orbit as it were. We don't have a big endian host to validate it on and I'd rather not waste any time on use-case that currently have zero users.

kamathba replied:
agreed.

kamathba commented on hw/net/hercules_emac.c:
```diff
@@ -0,0 +1,650 @@
+        dma_memory_read(&address_space_memory, s->txhdp[idx], &txd,
+                        sizeof(txd));
+
+        addr = txd.buffer_pointer;
```
do these all need le16toh wrappers?

kamathba commented on hw/net/hercules_emac.c:
```diff
@@ -0,0 +1,650 @@
+        dma_memory_write(&address_space_memory, s->txhdp[idx], &txd,
+                         sizeof(txd));
+
+        s->txhdp[idx] = txd.next;
```
le32toh?

kamathba commented on hw/net/hercules_emac.c:
```diff
@@ -0,0 +1,650 @@
+                                                const uint8_t *buf,
+                                                size_t size)
+{
+    HerculesCppiDescriptor rxd;
```
same kind of question for all of these re: le16toh

kamathba commented on hw/arm/hercules.c:
```diff
@@ -672,3 +676,75 @@ static void hercules_register_types(void)
+    qdev_prop_set_drive(DEVICE(dev), "eeprom",
+                        eeprom ? blk_by_legacy_dinfo(eeprom) : NULL,
+                        &error_abort);
+    object_property_set_bool(dev, is_tms570, "is_tms570", &error_fatal);
```
regarding instance_init vs realize, is this appropriate?

ndreys replied:
I think so, but I am also sure we'd get a lecture from upstream if it isn't :-) I don't know why, but for some reason I think property names are supposed to use `-` instead of `_`, so we might want to call it `is-tms570`

kamathba replied:
got it

kamathba replied:
Figured out we can just use object_new_with_props instead

kamathba commented on hw/arm/hercules.c:
```diff
@@ -206,6 +201,14 @@ static void hercules_realize(DeviceState *dev, Error **errp)
     qemu_irq irq, error;
     int i;
 
+    s->cpu.ctr = 0x1d192992; /* 32K icache 32K dcache */
```
I thought this was ok but just want to highlight it.

kamathba commented on hw/arm/hercules.c:
```diff
@@ -0,0 +1,674 @@
+         * FIXME: Convenience workaround, should eventually be removed
+         * once loading from custom file is implemented
+         */
+        const HerculesOTPData dummy_otp = {
```
This feels like it will be an irritating thing to deal with regarding LE/BE

ndreys replied:
Oh, this thing. When we were working on it, I eventually re-implemented it via `-device loader,file=<blah>,address=<blah>'` and plumbed it in glorious `durculess`, but you wanted to keep this dummy one for the sake of convenience. I think we should just drop this bit of code and rely on `-device loader`

kamathba replied:
Oh, absolutely. I had forgotten the durcules launcher was overriding this anyways

kamathba commented on hw/arm/hercules.c:
```diff
@@ -0,0 +1,674 @@
+
+            .die_temp_sensor = {
+                [0] = {
+                    .temp1    = htobe16(0x012F),
```
In particular all these

ndreys commented on accel/kvm/kvm-all.c:
```diff
@@ -2171,11 +2171,12 @@ static void kvm_handle_io(uint16_t port, MemTxAttrs attrs, void *data, int direc
 {
```
As discussed, let's drop this commit for now. We can fix debug access in a separate submission

ndreys commented on hw/intc/hercules_vim.c:
```diff
@@ -0,0 +1,314 @@
+    /*
+     * FIXME: Is this the best way to deal with endianness RM57 vs
+     * TMS570
+     */
```
@kamathba we should tests this one on RM57

ndreys commented on hw/ssi/hercules_spi.c:
```diff
@@ -0,0 +1,723 @@
+/*
+ *  Current assumptions in this MIBSPI implementation
+ *
+ *   - Only MIBSPI mode is implemented, no compatability mode
```
I don't think this is true anymore. I think I had to implement compatibility mode too. Right?

kamathba replied:
Yea, we use it on mibspi5 + dma

ndreys commented on hw/ssi/hercules_spi.c:
```diff
@@ -0,0 +1,723 @@
+static void hercules_spi_ram_write(HerculesMibSpiState *s, uint32_t *word,
+                                   uint32_t value)
+{
+    if (hercules_spi_ram_big_endian(s)) {
```
Another spot that we need to check on RM57

kamathba replied:
yea... I dug into this a big more and will comment separately

kamathba replied:
I at least checked that spi transfers complete in an expected way without the external chardev doing much (you can setup/start transfers)

ndreys commented on hw/ssi/hercules_spi.c:
```diff
@@ -0,0 +1,723 @@
+{
+    HerculesMibSpiState *s = opaque;
+
+    /* FIXME: This assumes BE */
```
Oh god, if only I could remember what this meant

Commit:
fixup! hw/arm: Add support for Hercules SoC

Commit:
hw/misc: Add support for Hercules ESM module

Signed-off-by: Andrey Smirnov <andrew.smirnov@gmail.com>
Signed-off-by: Benjamin Kamath <kamath.ben@gmail.com>

Commit:
fixup! hw/arm: Add support for Hercules SoC

Can probably make this a helper function to reduce some boilerplate.
What I would really want to do is allocate the MemoryRegionOps and
have it return a 'const' version, but that should probably happen in
init rather than realize...

kamathba commented:
So, the BE/LE problem with the hercules peripherals is pretty irritating. nothing outside of hw/arm/* has include access to target/arm/ (for cpu.h).

I think the way to do this is make "cpu" opaque in HerculesState, and like you said: check the parent object or global CPU objects endianness to set the MemoryRegionOps. I hate it.

ndreys commented on hw/adc/hercules_mibadc.c:
```diff
@@ -269,61 +270,70 @@ static void hercules_mibadc_write(void *opaque, hwaddr offset,
+    };
+
+    if (parent->is_tms570)
+    {
```
Dumbest of nits: here and in other if statements `{` is on a new line for some reason. Should be on the same line per coding style.
